src/Network/Transport/TCP.hs

```diff
@@ -928,34 +947,60 @@ handleConnectionRequest transport socketClosed (sock, sockAddr) = handle handleE
     mAddrInfo <- maybe (fmap Just) System.Timeout.timeout connTimeout $ do
       ourEndPointId <- recvWord32 sock
       let maxAddressLength = tcpMaxAddressLength $ transportParams transport
-      theirAddress <- EndPointAddress . BS.concat <$>
-        recvWithLength maxAddressLength sock
-      return (ourEndPointId, theirAddress)
-    (ourEndPointId, theirAddress) <- case mAddrInfo of
+      addressControlCode <- recvWord32 sock
+      case addressControlCode of
+        0 -> do
+          -- Let the peer tell us their EndPointAddress.
+          let maxAddressLength = tcpMaxAddressLength $ transportParams transport
+          theirAddress <- EndPointAddress . BS.concat <$>
+            recvWithLength maxAddressLength sock
+          (theirHost, theirPort, theirEndPointId)
+            <- maybe (throwIO (userError "handleConnectionRequest: peer gave malformed address"))
+                     return
+                     (decodeEndPointAddress theirAddress)
+          -- If the OS-determined host doesn't match the host that the peer gave us,
+          -- then we have no choice but to reject the connection. It's because we
+          -- use the EndPointAddress to key the remote end points (localConnections)
+          -- and we don't want to allow a peer to deny service to other peers by
+          -- claiming to have their host and port.
+          return (ourEndPointId, Right theirAddress)
+        -- The peer is not addressable, so we generate a random address which
+        -- will pick out this socket for its lifetime.
+        1 -> do
+          theirAddress <- randomEndPointAddress
```

Comment:
Shouldn't the peer chose the random address once before connecting? That way it would be easier to identify the node which doesn't have an external address.

Otherwise, what's the point of choosing a different address each time it connects?